Sources/PopupBridge/POPPopupBridge.swift

```diff
@@ -153,33 +148,32 @@ public class POPPopupBridge: NSObject, WKScriptMessageHandler, SFSafariViewContr
     
     // MARK: - WKScriptMessageHandler conformance
     
+    /// :nodoc: This method is not covered by Semantic Versioning. Do not use.
+    ///
+    /// Called when the webpage sends a JavaScript message back to the native app
     public func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
         if message.name == messageHandlerName {
-            guard let params = message.body as? [String: Any] else {
-                // TODO: - create error case & add unit test
+            guard let jsonData = try? JSONSerialization.data(withJSONObject: message.body),
+                  let script = try? JSONDecoder().decode(WebViewMessage.self, from: jsonData) else {
                 return
             }
             
-            if let urlString = params["url"] as? String,
+            if let urlString = script.url,
                let url = URL(string: urlString) {
                 dismissSafariViewController()
                 
                 delegate.popupBridge?(self, willOpenURL: url)
                 
                 let viewController = SFSafariViewController(url: url)
-                self.delegate.popupBridge(self, requestsPresentationOfViewController: viewController)
-                
                 safariViewController = viewController
                 safariViewController?.delegate = self
+                
+                self.delegate.popupBridge(self, requestsPresentationOfViewController: viewController)
+                return
+            } else if let name = script.message?.name {
+                delegate.popupBridge?(self, receivedMessage: name, data: script.message?.data)
```

Comment:
Do we think it's worth handling if `script.message.name` or `script.message.data` is missing? Since the message is optional I'm not totally sure what happens if the `name` or `data` is nil (or if that's fine).